Step-2: Add functionality to Add tasks

sgu-wtag opened the pull request:
[Step-2](https://trello.com/c/62dVpl5I/3-step-2-add-functionality-to-add-tasks)
This PR introduces adding tasks using an input field to the app store and showing tasks in bare-bone format on the task board.

## Types of changes
- [x] Add all the necessary input fields according to [design](https://www.figma.com/file/9Rwb2k9Dfbh8Ycb6kbJfzZ/Todo-App-UI-for-New-Joiner?node-id=0%3A1&t=fqQdwByuj6siGx5y-0)(Not the styling, just the elements themselves).
- [x] Connect store(Redux), and save tasks to store.
- [x] Show tasks in a bare-bone format in the task board.

## Review
Please have a close look at this PR and kindly put your valuable comments.

https://github.com/sse-wtag/todo-app-react/assets/144016889/0abcf5ab-272f-46d9-a3ac-4845dddaab78

sgu-wtag committed:
Add basic configuration for absolute import

sgu-wtag committed:
Add basic style configuration

sgu-wtag committed:
Add Navbar component

sgu-wtag committed:
Add necessary input fields according to design

sgu-wtag committed:
Add task addition functionality

sgu-wtag committed:
feat: Add input sanitization

sgu-wtag committed:
Refactor formatDate for improved readability & error handling

sgu-wtag committed:
Refactor UI elements

sgu-wtag committed:
Refactor: Convert Raw CSS to SCSS

iud-wtag commented on package.json:
```diff
@@ -12,6 +12,8 @@
     },
```
Best practice to provide engines. Which will give an overview of the stacks that used in project.
```suggestion
    },
    "engines": {
    "node": "16.x",
    "npm": "8.19.x",
    "yarn": "1.22.x"
  },
```

iud-wtag commented on src/components/Navbar.jsx:
```diff
@@ -0,0 +1,17 @@
+            <div className="nav__container">
+                <a href="#">
+                    <span>Todos</span>
+                </a>
```
Is there suppose to be link? I think we can remove the link and used the header as plain text as it is one page project.

iud-wtag commented on src/components/TaskBoard/index.jsx:
```diff
@@ -0,0 +1,22 @@
+import TaskBoardAction from "@components/TaskBoardAction";
+import Tasks from "@components/Tasks";
+import { useState } from "react";
+import "./style.scss";
```
Library file should be sort on top. and then local file.

iud-wtag commented on src/components/Tasks/TaskItem.jsx:
```diff
@@ -0,0 +1,20 @@
+}
+
+TaskItem.propTypes = {
+    task: PropTypes.object.isRequired,
```
consider defining todo object shape

iud-wtag commented on src/components/Tasks/TaskItem.jsx:
```diff
@@ -0,0 +1,20 @@
+import PropTypes from "prop-types";
+import "./style.scss";
+
+function TaskItem({ task = {} }) {
```
Why task = {} ?

sgu-wtag replied:
I’m using it as a default value for this `task` property. I think it is a good practice to add a default so that the app will never through an error while destructuring the property. I can even write `task || {}` to avoid crashing the app.
Please suggest to me the best practice here.
Thank you.

iud-wtag replied:
As we discussed, You can use `defaultProps` here for default values.

iud-wtag commented on src/components/Tasks/TaskItemInput.jsx:
```diff
@@ -0,0 +1,51 @@
+
+    return (
+        <form className="task-list__item" onSubmit={handleSubmit}>
+            <TextArea value={title} onChange={(e) => setTitle(e.target.value)} autoFocus required />
```
Can we use the arrow function in separate function and call it here?

iud-wtag commented on src/components/Tasks/style.scss:
```diff
@@ -0,0 +1,28 @@
+        border: 1px solid #9ca3af;
+        min-height: 10rem;
+
+        h2 {
+            font-size: 1.2rem;
+            font-weight: bold;
+            line-height: 2rem;
+        }
+
+        input {
+            width: 100%;
+            height: 7rem;
+            margin-bottom: 1rem;
+            text-align: start;
+        }
```
Consider using classnames instead of direct tag.

iud-wtag commented on src/components/ui/SearchBox/index.jsx:
```diff
@@ -0,0 +1,11 @@
+function SearchBox() {
+    return (
+        <form>
+            <TextInput placeholder="Search..." />
```
Can we use constant for the text?

iud-wtag commented on src/components/ui/form/TextArea/style.scss:
```diff
@@ -0,0 +1,6 @@
+    width: 100%;
+    border-radius: 0.5rem;
+    padding: 0.5rem;
+    background-color: #f9fafb;
```
Can we use color variable here?

sgu-wtag replied:
I've used colour variable here. Kindly review it.

iud-wtag commented on src/components/ui/form/TextInput/style.scss:
```diff
@@ -0,0 +1,10 @@
+input {
+    background-color: rgb(249 250 251);
+    border: 1px solid rgb(209 213 219);
```
Noticed that you have used rem. So can we be consistent here also with same format?

sgu-wtag replied:
I've used `px` here just because I don't want it to be changed based on the root. I want the `border-width` to be fixed. Please suggest me the best practice here.

iud-wtag commented on src/components/ui/form/TextInput/style.scss:
```diff
@@ -0,0 +1,10 @@
+input {
+    background-color: rgb(249 250 251);
```
use color variable with same format

iud-wtag commented on src/components/ui/form/button/index.jsx:
```diff
@@ -0,0 +1,18 @@
+}
+
+Button.propTypes = {
+    children: PropTypes.node.isRequired || PropTypes.string.isRequired,
```
What could we expect inside node?

sgu-wtag replied:
I've changed it and defined it with `PropTypes.element`.

iud-wtag commented on src/components/ui/form/button/style.scss:
```diff
@@ -0,0 +1,14 @@
+.btn {
+    align-items: center;
+    background-color: #fff;
+    border: 1px solid #9ca3af;
+    border-radius: 0.5rem;
+    color: #32394b;
+    display: inline-flex;
+    font-size: 0.875rem;
+    font-weight: bold;
+    gap: 0.5rem;
+    line-height: 1.25rem;
+    padding: 0.625rem 1.25rem 0.625rem 1.25rem;
+    text-align: center;
+}
```
ditto

sgu-wtag committed:
Update package.json

Provided engines in `package.json`

Co-authored-by: Imtiaj Rafi <124483067+iud-wtag@users.noreply.github.com>

sgu-wtag committed:
Remove link tag from Navbar

sgu-wtag committed:
Refactor: Separate title state handling

In this commit, I've refactored the task form by using a dedicated function to handle the `title` state updates in response to user input. This change improves code organization and readability by isolating the title state management logic within the `onChange` prop, enhancing maintainability and clarity in the component.

sgu-wtag committed:
Refactor: Update PropType for Improved Validation

In this commit, I've updated the PropType for a specific component to enhance validation and ensure better data integrity.

sgu-wtag committed:
Refactor: Define todo object shape

In this commit, I've defined todo object shape for better readability and maintainability.

sgu-wtag committed:
Refactor: Sort import order

sgu-wtag committed:
Refactor: Use color variable in CSS

sgu-wtag committed:
Refactor: Use constant for placeholder text

iud-wtag commented on src/components/Tasks/TaskItemInput.jsx:
```diff
@@ -0,0 +1,61 @@
+            return;
+        }
+
+        const newTask = {
+            id: Date.now().toString(),
+            title: cleanTitle,
+            createdAt: new Date().toISOString(),
+            isCompleted: false,
+        };
```
Can we move this part into separate redux component?

iud-wtag commented on src/components/Tasks/style.scss:
```diff
@@ -0,0 +1,31 @@
+        background-color: base.$bg-white;
+        padding: 1rem;
+        border-radius: 0.5rem;
+        border: 1px solid base.$bg-gray-400;
```
please take care of this.

iud-wtag commented on src/components/ui/SearchBox/index.jsx:
```diff
@@ -0,0 +1,13 @@
+import TextInput from "@components/ui/form/TextInput";
+
+function SearchBox() {
+    const PLACEHOLDER = "Search...";
```
Are we using any separate file for managing all constants? Is so then move it accordingly. Otherwise it is good practice to store all constant in one separate helper file.

sgu-wtag replied:
But here, this placeholder text will always be used only for this specific component. So, this is just a local constant variable. After reviewing this code, I have come up with the idea that I should make this `placeholder` a `prop` for this `SearchBox` component and give it a default value of “search...“.
Kindly review it and suggest the best practice here.
Thank you.

iud-wtag commented on src/components/ui/form/TextArea/index.jsx:
```diff
@@ -0,0 +1,13 @@
+import "./style.scss";
+
+function TextArea({ className, rows = 3, ...rest }) {
+    return <textarea className={className || "input-textarea"} rows={rows} {...rest}></textarea>;
```
Can we introduce `classnames` here?

https://www.npmjs.com/package/classnames

iud-wtag replied:
Resolved.

iud-wtag commented on src/components/ui/form/TextInput/style.scss:
```diff
@@ -0,0 +1,12 @@
+
+input {
+    background-color: base.$bg-gray-50;
+    border: 1px solid base.$bg-gray-300;
```
ditto

iud-wtag commented on src/components/ui/form/TextInput/style.scss:
```diff
@@ -0,0 +1,12 @@
+    background-color: base.$bg-gray-50;
+    border: 1px solid base.$bg-gray-300;
+    border-radius: 0.5rem;
+    display: block;
```
is `display: block` necessary here?

sgu-wtag replied:
No, I've removed it. Thank you.

iud-wtag commented on src/components/ui/form/button/index.jsx:
```diff
@@ -0,0 +1,18 @@
+}
+
+Button.propTypes = {
+    children: PropTypes.oneOfType([PropTypes.string, PropTypes.element]).isRequired,
```
Good work :tada:

iud-wtag commented on src/components/ui/form/button/style.scss:
```diff
@@ -0,0 +1,16 @@
+.btn {
+    align-items: center;
+    background-color: base.$bg-white;
+    border: 1px solid base.$bg-gray-400;
```
ditto

iud-wtag commented on src/components/ui/form/button/style.scss:
```diff
@@ -0,0 +1,16 @@
+    border: 1px solid base.$bg-gray-400;
+    border-radius: 0.5rem;
+    color: base.$bg-slate-700;
+    display: inline-flex;
```
whats the purpose of `inline` here?

sgu-wtag replied:
Here, I want the button element to flow `inline` with the text content. Thank you.

iud-wtag replied:
can't we make this with `flex` only?

sgu-wtag replied:
Changed. Thank you.

iud-wtag commented on src/styles/_base.scss:
:tada:

iud-wtag commented on src/styles/navbar.scss:
```diff
@@ -0,0 +1,22 @@
+
+.nav {
+    background-color: base.$bg-white;
+    border-bottom: 1px solid base.$secondary-color;
```
ditto

sgu-wtag committed:
Refactor: Use `defaultProps` for default values

sgu-wtag committed:
Fix: `formatDate` file import error

sgu-wtag committed:
Refactor: Change spacing from `px` to `rem`

sgu-wtag committed:
Refactor: Remove unnecessary `display: block`

sgu-wtag committed:
Refactor: Follow proper BEM methodology

sgu-wtag committed:
Refactor: Use classnames instead of direct tag

sgu-wtag committed:
Refactor: Add `defaultProps` and base `btn` class

sgu-wtag committed:
Style: Basic textarea

fza-wtag commented on src/components/Navbar.jsx:
```diff
@@ -0,0 +1,15 @@
+function Navbar() {
+    return (
+        <nav className="nav">
+            <div className="nav__container">
```
Bhai, which naming convention we are following here?

sgu-wtag replied:
I'm following the `BEM` naming convention here, @fza-wtag bhai.

fza-wtag replied:
Cool! Thanks.

iud-wtag reviewed (changes requested):
Can you please confirm the following condition meeting?

- [x] Empty task or only spaces should not be added.
- [x] Sanitizing scripts and tag.
- [x] No Leading and Trailing spaces.
- [x] Task should added on board when press `Enter` in keyboard(Also make sure all validations).

Check the boxes if condition met.

sgu-wtag committed:
Feat: Add task on board when press `Enter`

sgu-wtag committed:
Refactor: Change display mode for button

iud-wtag commented on src/components/Tasks/TaskItemInput.jsx:
```diff
@@ -0,0 +1,69 @@
+    };
+
+    const handleTextareaKeyDown = (event) => {
+        if (event.key === "Enter" && !event.shiftKey) {
```
Constantssssss... :eyes:

iud-wtag reviewed (approved):
Good Work. LGTM :tada:

sgu-wtag committed:
Refactor: Move constants in a separate file

sgu-wtag committed:
Refactor: Re-structure folders

sgu-wtag committed:
Refactor: Move task card styles inside its folder

sgu-wtag committed:
Fix: Unexpected `undefined` class is getting added

ray-wtag commented on src/components/TaskBoard/TaskBoard.jsx:
```diff
@@ -0,0 +1,30 @@
+            <div className="task-board__action-wrapper">
+                <Button onClick={toggleCreation}>Create</Button>
+                <div className="task-board__filter-buttons">
+                    <Button>all</Button>
```
```suggestion
                    <Button>All</Button>
```

sgu-wtag replied:
@ray-wtag bhaia, I've kept the text all lower case intentionally so that I can control it using CSS. And I've CSS property `capitalize` here. Please, suggest me what is the best practice here...

ray-wtag replied:
it seems like you are taking the long road. It doesn't need any CSS, where you can just write `All` instead of `all`.

ray-wtag commented on src/components/TaskBoard/TaskBoard.jsx:
```diff
@@ -0,0 +1,30 @@
+                <Button onClick={toggleCreation}>Create</Button>
+                <div className="task-board__filter-buttons">
+                    <Button>all</Button>
+                    <Button>incomplete</Button>
```
```suggestion
                    <Button>Incomplete</Button>
```

sgu-wtag replied:
@ray-wtag bhaia, same here...

ray-wtag commented on src/components/TaskBoard/TaskBoard.jsx:
```diff
@@ -0,0 +1,30 @@
+                <div className="task-board__filter-buttons">
+                    <Button>all</Button>
+                    <Button>incomplete</Button>
+                    <Button>complete</Button>
```
```suggestion
                    <Button>Complete</Button>
```

sgu-wtag replied:
@ray-wtag bhaia, same as well...

ray-wtag commented on src/components/TaskList/TaskList.jsx:
```diff
@@ -0,0 +1,27 @@
+import "./TaskList.scss";
+
+function TaskList({ isTaskCreating, onTaskCreation }) {
+    const tasks = useSelector((state) => state.tasks.tasks);
```
can't we use any other name here? `state.tasks.tasks` looking a bit confusing

sgu-wtag replied:
@ray-wtag bhaia, I've changed it from `state.tasks.tasks` to `state.todo.tasks`. Thanks!

ray-wtag commented on src/components/ui/form/Button/Button.jsx:
```diff
@@ -0,0 +1,30 @@
+        <button
+            type={type}
+            className={classNames({
+                btn: true,
```
if `btn` is always true, you don't have to use it in any condition. You can write it like
className={classNames('btn'{.....}

ray-wtag commented on src/components/ui/form/TextInput/TextInput.jsx:
```diff
@@ -0,0 +1,18 @@
+import PropTypes from "prop-types";
+import "./TextInput.scss";
+
+function TextInput({ className, type, ...rest }) {
```
It may confuse with `TextArea`. We can just say it `Input`

sgu-wtag replied:
I've changed it to `Input`. Thanks!

ray-wtag commented on src/features/task/taskSlice.js:
```diff
@@ -7,7 +7,12 @@ const initialState = {
-    reducers: {},
+    reducers: {
+        addTask: (state, action) => {
+            state.tasks.unshift(action.payload);
```
using `unshift` for showing recent tasks at the top right?

sgu-wtag replied:
yes, bhaia. So that I don't need to reverse it always to show the recent tasks at the top. Thank you!

sgu-wtag committed:
Refactor: Change component `TextInput` to `Input`

sgu-wtag committed:
Refactor: Change button's default class

sgu-wtag committed:
Refactor: Change state name from `tasks` to `todo`

sgu-wtag committed:
Fix: Button is showing error when adding multiple elements

sgu-wtag committed:
Fix default class & remove unnecessary `rows` prop

sgu-wtag committed:
Fix: created date is showing full year

ray-wtag commented on src/styles/_base.scss:
maybe we can name it `_variables.scss`

ray-wtag commented on src/helpers/formatting/formatDate.js:
```diff
@@ -0,0 +1,25 @@
+
+        const dateObject = new Date(date);
+
+        if (isNaN(dateObject.getTime())) {
```
maybe we don't have to do this, we can just check whether dateObj is undefined or not

ray-wtag commented on src/helpers/formatting/formatDate.js:
```diff
@@ -0,0 +1,25 @@
+        }
+
+        const options = { day, month, year };
+        const formattedDate = dateObject.toLocaleDateString(locale, options).replace(/\//g, separator);
```
can't we format without using regex?

ray-wtag commented on src/helpers/text/purify.js:
```diff
@@ -0,0 +1,17 @@
+    }
+
+    try {
+        const purifiedText = DOMPurify.sanitize(dirtyText, { USE_PROFILES: useProfiles });
```
what's the use of useProfiles here?

sgu-wtag replied:
Here, the `useProfiles` parameter offers precise control over DOMPurify's HTML handling. When `html` is `true`, it allows safe HTML elements and attributes while removing malicious content, resulting in purified HTML. On the other hand, when `html` is set to `false`, it treats the input as plain text, removing HTML tags and attributes to produce plain text output.

sgu-wtag committed:
Refactor: Rename file for clarity

sgu-wtag committed:
Refactor: Simplify and clarify date formatter

sgu-wtag committed:
Refactor: Capitalize filter button text

sgu-wtag committed:
Refactor: Update variable import for improved SCSS

fza-wtag commented on src/components/TaskCard/TaskCard.jsx:
```diff
@@ -0,0 +1,27 @@
+    );
+}
+
+TaskCard.defaultProps = {
+    task: {},
+};
+
+TaskCard.propTypes = {
+    task: PropTypes.shape({
+        title: PropTypes.string,
+        createdAt: PropTypes.string,
+    }),
+};
```
I think there should not be any use case where we call `TaskCard` without `task` as prop. Thus me may consider doing this change?

```suggestion
TaskCard.propTypes = {
    task: PropTypes.shape({
        title: PropTypes.string,
        createdAt: PropTypes.string,
    }).isRequired,
};
```

fza-wtag reviewed (approved):
Nice work! 👏

sgu-wtag committed:
Update src/components/TaskCard/TaskCard.jsx

Co-authored-by: Ferdaus Zaman Polok <119289799+fza-wtag@users.noreply.github.com>
